### ISSUE TITLE
As an administrator, I can deploy an application image using the bootable jar with the WildFly Operator

### ISSUE THREAD

**Issue by yersan** [wildfly/wildfly-operator#166]
## Overview
The EAP Operator input is an application image. Until now, the requirement was that this application image was built using EAP S2I builder and runtime images. Since WildFly 21, the Bootable JAR (also known as “fat JAR”) is a new way to package a WildFly server and an application both in an executable JAR. 

The goal of this RFE is to add the ability to deploy a Bootable JAR application image by using the WildFly Operator.

## Dev Contacts
Yeray Borges (yborgess@redhat.com)
## QE Contacts
Jan Blizňák (jbliznak@redhat.com)
## Testing By
QE
## Affected Projects or Components
[WildFly Operator](https://github.com/wildfly/wildfly-operator)
## Other Interested Projects
N/A
### Hard Requirements

- Bootable JAR application image requirements:
   - The base image for the Bootable JAR container must be registry.access.redhat.com/ubi8/openjdk-11 or any other Red Hat ubi8 supplying a higher JDK version.
   - The Bootable JAR must be built for a server configured for the cloud environment. This is done by enabling the <cloud> configuration item on the wildfly-jar-maven-plugin configuration. See https://docs.wildfly.org/bootablejar/#wildfly_jar_configuring_cloud
   - For example, one way to generate a Bootable JAR application image from an existing Bootable JAR project is by using the [JKube Kubernetes maven plugin ](https://www.eclipse.org/jkube/docs/kubernetes-maven-plugin) . The following is an example of its uses in a maven project:
 ```
<project xmlns="http://maven.apache.org/POM/4.0.0" ...>
   ...
    <properties>
           ...
        <jkube.generator.from>registry.redhat.io/ubi8/openjdk-11:latest</jkube.generator.from>
    </properties>

    <dependencies>
        ...
    </dependencies>

    <build>
        <finalName>${project.artifactId}</finalName>
        <plugins>
            ...
            <plugin>
                <groupId>org.wildfly.plugins</groupId>
                <artifactId>wildfly-jar-maven-plugin</artifactId>
                <version>${version.bootable.jar}</version>
                <configuration>
                    <feature-pack-location>wildfly@maven(org.jboss.universe:community-universe)#${version.wildfly}</feature-pack-location>
                    <layers>
                        <layer>...</layer>
                    </layers>
                    <cloud/>
                </configuration>
                <executions>
                    <execution>
                        <goals>
                            <goal>package</goal>
                        </goals>
                    </execution>
                </executions>
            </plugin>
            <plugin>
                <groupId>org.eclipse.jkube</groupId>
                <artifactId>kubernetes-maven-plugin</artifactId>
                <version>${version.jkube}</version>
            </plugin>
        </plugins>
    </build>
</project>
```
Upon the above pom.xml example, the Application Bootable JAR image can be generated by issuing the following maven goals:
`mvn package k8s:build`
The `kubernetes-maven-plugin` will create a Dockerfile available at `${project.build.directory}/target/docker`. This image is also tagged and made available on your local docker registry.
- Custom Resource configuration requirements
   - The “applicationImage” attribute will be used to identify the Bootable JAR application image. It will accept the same formats already available for S2I application images:
     - An image name: quay.io/bootable-jar/wildfly-demo
     - An image tag: quay.io/bootable-jar/wildfly-demo:1.0
     - An image hash: quay.io/bootable-jar/wildfly-demo@sha256:47c06c96e80d0defb777686cdb468c636d9b3b7081a35b784330a050a403e15b
     - An OpenShift image stream: wildfly-demo:latest
   - The WildFly operator CRD will supply a new boolean configuration attribute named “bootable-jar”, defaulted to false, to identify whether the application image is a Bootable JAR or if it is a S2I application image.
      - This value must be ‘true’ to enable the Bootable JAR images.
- Transaction Recovery facility is not supported on the Bootable JAR. 
    - The WildFly Operator will not run the Transaction Recovery facility when a pod running a Bootable JAR application image is scaled down.
    - The WildFly Operator will log a trace describing this situation when a pod is scaled down.
- Supply custom server configuration file via config map is not supported a Bootable JAR application image.
    - The natural way to proceed with a Bootable JAR is to configure the server via CLI script when the Bootable JAR is created.
- The WildFly Operator will inject the liveness and readiness probes. The injected probes will be:
   - liveness: An HTTPGet probe that checks the /health/live resource
   - readiness: An HTTPGet probe that checks /health/ready resource

### Nice-to-Have Requirements
N/A

### Non-Requirements
N/A

## Test Plan
- An example Custom Resource using a Bootable JAR image will be supplied under the deploy/crds directory. This example CR can be used to run a manual test and verify the Operator deploys the image in an OpenShift cluster.
- The Operator test suite will include a test case which will deploy a Bootable JAR application image to verify the required replicas are created without errors.

## Community Documentation
The user guide documentation and the WildFlyServer CRD documentation will involve to reflect the changes introduced by this RFE

## Release Note Content
Added the ability to deploy a Bootable JAR image.

**Comment by jmesnil** [comment]
@yersan that looks good to me.

We should make sure that the transaction recovery issue is properly mentioned in the release notes so that our users are not surprised by it.

Coud you also please mention that the operation will inject the *HTTP* liveness and readiness probes (to be more explicit).

**Comment by mchoma** [comment]
- > The base image for the Bootable JAR container must be registry.access.redhat.com/ubi8/openjdk-11  
  - I wonder Where does this requirement come from? Is this somewhere explicitely stated on bootable jar requirement. XP2 is still supported on jdk8 images and therefore on this option we are using openjdk-8 images. Is that wrong? Is it safe to use just registry.access.redhat.com/ubi8/openjdk-11 ?
- > new boolean configuration attribute named “bootable-jar”
  - Is this really necessary? Can't that be detected from image? What is handleded by this flag? Transaction recovery and http probes? Can't these flags be named by specific feature they are handling?
- > The WildFly Operator will inject the liveness and readiness probes.
  - what exactly that mean. Instead of shell probes http will be used? Given http probes are present since released 7.3.4 (and customers should use latest released images), theoretically it should be safe to switch to http probes by default?

**Comment by jmesnil** [comment]
> > new boolean configuration attribute named “bootable-jar”
> 
> * Is this really necessary? Can't that be detected from image? What is handleded by this flag? Transaction recovery and http probes? Can't these flags be named by specific feature they are handling?

This attribute is necessary as we don't have a way to identify where the application image is a Bootable Jar or a S2I image.
There are differences between the 2 images that are captured by this flag including the location of the server (as the Bootable Jar image does not define a JBOSS_HOME).

> > The WildFly Operator will inject the liveness and readiness probes.
> * what exactly that mean. Instead of shell probes http will be used? Given http probes are present since released 7.3.4 (and
> customers should use latest released images), theoretically it should be safe to switch to http probes by default?

Our existing S2I images does not mandate the presence of the `health` subsystem and use shell probes to cope with its absence.
For bootable Jar, these shell probes are not an option (as they were provided by the S2I builder images) so the `health` subsystem presence becomes a requirement and HTTP probes are the only way to determine the pod healthiness

**Comment by mchoma** [comment]
> This attribute is necessary as we don't have a way to identify where the application image is a Bootable Jar or a S2I image.

Is it possible to load metadata from image itself? In java testsuite we load metadata through java API [1]. So I assume something similar could be possible with Go. Of course that could be less reliable than static parameter. But would mean no aditional parameter in API. What if in future there will be third type of image Operator will have to handle? There will be another flag?

[1] https://github.com/xtf-cz/xtf/blob/6c4baf6f6584ef94e08b7f7920f5103a85becce7/junit5/src/main/java/cz/xtf/junit5/model/DockerImageMetadata.java#L102

**Comment by jmesnil** [comment]
That's different.
The only thing we have is the string corresponding to the application image. The operator does not have access to the image itself. It might not be pulled before the pod are created.

**Comment by yersan** [comment]
Hello @jfdenise, we are discussing the minimal requirements for the Bootable JAR. My understanding so far was it needs Java 11 as the minimal JDK version. @mchoma asked about this requirement, could you clarify here his questions? Thanks!

> I wonder Where does this requirement come from? Is this somewhere explicitely stated on bootable jar requirement. XP2 is still supported on jdk8 images and therefore on this option we are using openjdk-8 images. Is that wrong? Is it safe to use just registry.access.redhat.com/ubi8/openjdk-11 ?

**Comment by jfdenise** [comment]
FYI: For wildfly, we only have a jdk11 based wildfly-s2i image.
The upstream examples for Bootable JAR are only using JDK11 image, JDK8 has never been used nor mentioned.

**Comment by yersan** [comment]
Then, from the @jfdenise input, we have been testing and using only jdk1, indeed it is the current LTS release. Bootable JAR could work on JDK8 but that is not our target release, so there is no guaranty it works always on JDK 8.

@mchoma Does it clear up all your concerns?

**Comment by mchoma** [comment]
My concern was in context of EAP. I can't probably link that document here `EAP7-1385, Bootable jar, XP 2.0 specifics` but here it was mentioned bootable jar is supported on OpenJDK8 image as well. After discussion with Bilge, OpenJDK8 image was marked as deprecated.

**Comment by mchoma** [comment]
I would like to react to https://github.com/wildfly/wildfly-operator/issues/166#issuecomment-755221431. (github issues is not good platform for discussions :))

Is it possible to start pod and find out from there if it is bootable jar image or non-bootable jar image?

**Comment by jbliznak** [comment]
> The only thing we have is the string corresponding to the application image. The operator does not have access to the image itself. It might not be pulled before the pod are created.

Is the last two sentences really true? I don't pretend I have a knowledge of Operator (SDK) code or design but just thinking quickly I don't understand why it shouldn't be possible. What was Martin asking before seems to me that we would like to leverage something like `docker inspect `, not sure if that is possible without actually pulling the image, but we still need the image for starting the pods so what is preventing us to "preload" it to internal registry if not already there?

Note: I acknowledge the bootableJar param is working and easy to implement solution, however, UX experience is very big drawback so we are trying to ensure we really ruled out alternative ways before making this change "permanent"

**Comment by jmesnil** [comment]
this is a good suggestion and something we will look into when we want to have more metering from the operator to get more insights on deployed applications.
Given our timeframe, this is not achievable to do it for the upcoming release.

If and when we would be able to determine the type of application image from the EAP Operator, this would not prevent to deprecate the `bootableJar` field and make it a no-op.
Once we have migration webhook in place, we will also be able to remove it altogether without breaking anything.

**Comment by mchoma** [comment]
> Once we have migration webhook in place, we will also be able to remove it altogether without breaking anything.

I want to ask on timefrime of all actions needed. So ideally with Operator 3.0 [1] we will have a way how to get rid of  bootable-jar attribute. In that time we could be ready for using migration webhook and finish [2]. And also with that we can get rid of bootable-jar attribute as well? 

[1] https://issues.redhat.com/issues/?jql=project%3D%22EAP7%22%20AND%20%22Target%20Release%22%3D%22OP-3.0.0.GA%22%20ORDER%20BY%20priority%20ASC
[2] https://issues.redhat.com/browse/EAP7-1457

**Comment by yersan** [comment]
With a conversion webhook we should be able to convert "legacy" custom resources into a new format, so, we should be able to get rid of attributes if they are unneeded. Notice that conversion webhooks are supported by OLM only since OpenShift 4.6. We were able to have a POC for OpenShift 4.5 though, and found a minor issue on OpenShift 4.4 where some webhook resources should be removed manually when the Operator was uninstalled because the shipped OLM version on OpenShift 4.4 is unable to remove some resources automatically. Maybe there could be a workaround to fix it. Webhook POC investigation was stopped in favor of other priorities.

I'm not sure when or which OpenShift versions must be supported by Operator 3.0.0.GA yet, my current understanding is depending on which version, we still need to decide if the WebHook POC is enough for them so we can use it until we are able to use OLM directly to manage the webhooks for all the OpenShift supported versions.

**Comment by mchoma** [comment]
My understanding is we say we support 2 latest OCP 4 version. Currently it is 4.6 and 4.5.  Once 4.7 will be released in February we will support 4.7 and 4.6.
